### ISSUE TITLE
Support `CI_JOB_TOKEN` auth

### ISSUE THREAD

**Issue by dosuken123** [semantic-release/gitlab#156]
GitLab CI injects `CI_JOB_TOKEN` to allow a pipeline job to access a project resource through public v4 API. Not all of APIs support job token auth, however, [release creation API](https://docs.gitlab.com/ee/api/releases/#create-a-release) is already supported. So in theory, executing `GITLAB_TOKEN=$CI_JOB_TOKEN` should be suffice to run semantic-release.

However, as I tested https://gitlab.com/dosuken123/semantic-release-test/-/jobs/628942282, it didn't succeed as semantic-release tried [remote Git repository authentication](https://github.com/semantic-release/gitlab#gitlab-authentication). I'm not sure why it's necessary as the tag creation is done via the above API call.

If `CI_JOB_TOKEN` auth is possible, it's very convenient to run semantic-release in GitLab CI as users don't need to create [PAT](https://github.com/semantic-release/gitlab#gitlab-authentication).

**Comment by mmuenker** [comment]
For me, the `CI_JOB_TOKEN` works fine.

**GitLab**: 13.4.0-pre
**GitLab Runner**: 13.3.1 (shared runner)

**.gitlab-ci.yml**
```yaml
stages:
- release

release:
  needs: []
  image: registry.gitlab.com/rxap/gitlab-ci/semantic-release
  stage: release
  script: semantic-release
  variables:
    GITLAB_TOKEN: $CI_JOB_TOKEN
  interruptible: true
  rules:
  - if: '$CI_COMMIT_REF_NAME == "master"'
    when: on_success
```

**Comment by mattkasa** [comment]
@mmuenker I'm not able to reproduce a working release using your example with `$CI_JOB_TOKEN`, I get:

```
[9:15:28 PM] [semantic-release] › ✖  EGITNOPERMISSION Cannot push to the Git repository.
```

Can you think of anything different in your set up that would cause it to work?

**Comment by virtuman** [comment]
is it possible that it was meant to work in gitlab 13.4.* and not 13.3 ?

**Comment by EdwinSmulders** [comment]
I've tried this but I can't find any circumstances that made this work like @mmuenker says is possible. However, I also cannot find why it doesn't work - it seems like it fails before the gitlab plugin even does anything.

**Comment by secustor** [comment]
I have looked into this. At the moment I don't see how this could work at the moment, at least for `/api/v4/`.  
Option to deactivate the testing of git pushes as CI tokens have not the permission to do these. See [here](https://docs.gitlab.com/ee/api/README.html#gitlab-ci-job-token).     
The issue would need to be addressed in the sematic-release core. This is the line which results in the error. https://github.com/semantic-release/semantic-release/blob/master/index.js#L79

Could be that other rights are missing too.

**Comment by MRigal** [comment]
I can confirm this has no chance to work as @secustor said. The error message is 
```
'EGITNOPERMISSION',
  details: '**semantic-release** cannot push the version tag to the branch `semantic-release` on the remote Git repository with URL `blablabla`.\n' +
  ```

I don't think git push should be required here and being able to use the CI_JOB_TOKEN would ease a lot the usage of semantic-release!

**Comment by dosuken123** [comment]
It seems failing at [`verifyAuth`](https://github.com/semantic-release/semantic-release/blob/master/index.js#L77) in [semantic-release](https://github.com/semantic-release/semantic-release) (not `semantic-release-gitlab` module).

It's kinda make sense to auth if the token has a write permission to the repository, however, the [release creation API](https://docs.gitlab.com/ee/api/releases/#create-a-release) technically doesn't require the write permission to create a git-tag in its subsequent process.

**Comment by dosuken123** [comment]
Alternatively,  maybe we should use [Deploy Tokens](https://docs.gitlab.com/ee/user/project/deploy_tokens/) instead. This is more appropriate than Personal Access Token. However, there are no options to give write repository permission in the deploy token ATM, see https://gitlab.com/gitlab-org/gitlab/-/issues/23067.

**Comment by MRigal** [comment]
Yes, the deploy and ci tokens are pretty limited (and for good reasons by default) The issue you reference have moved to this epic now: https://gitlab.com/groups/gitlab-org/-/epics/3559

Would you like to create a feature request on `semantic-release` to be able to have a `verifyAuth(strict=false)` or similar since indeed gitlab does not need write permissions to create the new tag?

**Comment by nfriend** [comment]
[This thread](https://github.com/semantic-release/semantic-release/issues/1729) helped me better understand why the `CI_JOB_TOKEN` fails the `verifyAuth` step, even though the token _does_ give permission to access [GitLab's Tags API](https://docs.gitlab.com/ee/api/tags.html#create-a-new-tag).

**tl;dr:** `semantic-release` creates tags directly through Git, not through the GitLab API, so the token needs to have repository write access (which `CI_JOB_TOKEN` does not have).

**Comment by trietsch** [comment]
Does the OP still have issues with this? I've just tried what @mmuenker showed as an example, and for me, it's working as it should. A new release is created in GitLab.

**Comment by trietsch** [comment]
@dosuken123 do you still have issues with this? I've just tried what @mmuenker showed as an example, and for me, it's working as it should. A new release is created in GitLab.

**Comment by AljoschaP** [comment]
Is someone actively working on it?

**Comment by dosuken123** [comment]
Ah cool. Then, we can close this issue :)

**Comment by autotmp** [comment]
I'd imagine [gitlab-org/gitlab issue 223679](https://gitlab.com/gitlab-org/gitlab/-/issues/223679) should resolve this but the timeline seems bit fluid (and far out there).

**Comment by rene84** [comment]
We've made it work with CI_JOB_TOKEN by making just a few small changes to semantic-release and semant-release/gitlab. As described in another thread [here](https://github.com/semantic-release/semantic-release/issues/1729?notification_referrer_id=MDE4Ok5vdGlmaWNhdGlvblRocmVhZDE0NzA5NDAxNTQ6NzA5OTQ2OQ%3D%3D#issuecomment-1354472875)

**Comment by athlan** [comment]
I'd like to bring back an attention to that topic to use `CI_JOB_TOKEN`.

Current limitation is identified, that semantic-release core is doing a check whether it can push a tag. On GitLab integration, creating a tag is not needed. Creation of a Release (via api, what this project is doing) is sufficient enough to make a release, because it results with a tag, as mentioned above (https://github.com/semantic-release/gitlab/issues/156#issuecomment-747163220). What is more, `CI_JOB_TOKEN` can create a release, while it cannot push a tag to satisfy the semantic-release core lifecycle.

For that purpose I think there should be possible to skip pushing a tags by semantic-release core and then this check could be optional.

The rationale behind that is:
* CI_JOB_TOKEN are short-lived tokens with permissions derived from the triggerer. If ran by automation process they can access repo, but not modify the repo.
* Personal Access Tokens are too wide to be used - the leakage or extracting from another peer poses the token's issuer into a risk to impersonate changes on his behalf.
* Group level Access Tokens are not available in all organizations.
* Project level Access Tokens are posed as similar leakage as Personal Access Tokens.

Having that said, CI_JOB_TOKEN is the most secure way of triggering a Release process on GitLab.

**Comment by travi** [comment]
we hear the request. see https://github.com/semantic-release/semantic-release/issues/1729#issuecomment-1882385859 for our current response

**Comment by travi** [comment]
since this depends on changes in core to enable this, having two threads related to this is making this conversation more difficult to track. closing this in favor of https://github.com/semantic-release/semantic-release/issues/1729. let's please consolidate further conversation there

**Comment by fgreinacher** [comment]
> https://gitlab.com/gitlab-org/gitlab/-/issues/468320 This might be possible soon in Gitlab 17.3 or 17.4

Thanks for sharing @wwuck, I have not yet seen that!

With that functionality in place we would not need any `semantic-release` change. I'll therefore reopen this issue.

/cc @travi @JonasSchubert

**Comment by ReazerDev** [comment]
I enabled the feature flag for the feature @wwuck mentioned, and it does pass: `✔  Allowed to push to the Git repository`.
However it now fails with the following error:
```
[7:31:15 AM] [semantic-release] › ℹ  Start step "verifyConditions" of plugin "@semantic-release/gitlab"
[7:31:15 AM] [semantic-release] [@semantic-release/gitlab] › ℹ  Verify GitLab authentication (https://[GIT_URL]/api/v4)
[7:31:15 AM] [semantic-release] › ✘  Failed step "verifyConditions" of plugin "@semantic-release/gitlab"
[7:31:15 AM] [semantic-release] › ℹ  Start step "fail" of plugin "@semantic-release/gitlab"
[7:31:15 AM] [semantic-release] [@semantic-release/gitlab] › ℹ  Verify GitLab authentication (https://[GIT_URL]/api/v4)
[7:31:16 AM] [semantic-release] › ✘  Failed step "fail" of plugin "@semantic-release/gitlab"
[7:31:16 AM] [semantic-release] › ✘  EINVALIDGLTOKEN Invalid GitLab token.
The GitLab token (https://github.com/semantic-release/gitlab/blob/master/README.md#gitlab-authentication) configured in the GL_TOKEN or GITLAB_TOKEN environment variable must be a valid personal access token (https://docs.gitlab.com/ce/user/profile/personal_access_tokens.html) allowing to push to the repository [REPO].
Please make sure to set the GL_TOKEN or GITLAB_TOKEN environment variable in your CI with the exact value of the GitLab personal token.
[7:31:16 AM] [semantic-release] › ✘  EINVALIDGLTOKEN Invalid GitLab token.
The GitLab token (https://github.com/semantic-release/gitlab/blob/master/README.md#gitlab-authentication) configured in the GL_TOKEN or GITLAB_TOKEN environment variable must be a valid personal access token (https://docs.gitlab.com/ce/user/profile/personal_access_tokens.html) allowing to push to the repository [REPO].
Please make sure to set the GL_TOKEN or GITLAB_TOKEN environment variable in your CI with the exact value of the GitLab personal token.
AggregateError: 
    SemanticReleaseError: Invalid GitLab token.
```

**Comment by OrbisK** [comment]
> I enabled the feature flag for the feature @wwuck mentioned, and it does pass: `✔ Allowed to push to the Git repository`. However it now fails with the following error:
> 
> ```
> [7:31:15 AM] [semantic-release] › ℹ  Start step "verifyConditions" of plugin "@semantic-release/gitlab"
> [7:31:15 AM] [semantic-release] [@semantic-release/gitlab] › ℹ  Verify GitLab authentication (https://[GIT_URL]/api/v4)
> [7:31:15 AM] [semantic-release] › ✘  Failed step "verifyConditions" of plugin "@semantic-release/gitlab"
> [7:31:15 AM] [semantic-release] › ℹ  Start step "fail" of plugin "@semantic-release/gitlab"
> [7:31:15 AM] [semantic-release] [@semantic-release/gitlab] › ℹ  Verify GitLab authentication (https://[GIT_URL]/api/v4)
> [7:31:16 AM] [semantic-release] › ✘  Failed step "fail" of plugin "@semantic-release/gitlab"
> [7:31:16 AM] [semantic-release] › ✘  EINVALIDGLTOKEN Invalid GitLab token.
> The GitLab token (https://github.com/semantic-release/gitlab/blob/master/README.md#gitlab-authentication) configured in the GL_TOKEN or GITLAB_TOKEN environment variable must be a valid personal access token (https://docs.gitlab.com/ce/user/profile/personal_access_tokens.html) allowing to push to the repository [REPO].
> Please make sure to set the GL_TOKEN or GITLAB_TOKEN environment variable in your CI with the exact value of the GitLab personal token.
> [7:31:16 AM] [semantic-release] › ✘  EINVALIDGLTOKEN Invalid GitLab token.
> The GitLab token (https://github.com/semantic-release/gitlab/blob/master/README.md#gitlab-authentication) configured in the GL_TOKEN or GITLAB_TOKEN environment variable must be a valid personal access token (https://docs.gitlab.com/ce/user/profile/personal_access_tokens.html) allowing to push to the repository [REPO].
> Please make sure to set the GL_TOKEN or GITLAB_TOKEN environment variable in your CI with the exact value of the GitLab personal token.
> AggregateError: 
>     SemanticReleaseError: Invalid GitLab token.
> ```
> 
> From my inital research, it's because an api request is being made here: https://github.com/semantic-release/gitlab/blob/master/lib/verify.js#L67
> 
> Either the Job Token isn't allowed to make a request to the `/projects` endpoint at all (it's not listed here: https://docs.gitlab.com/ee/ci/jobs/ci_job_token.html) or it's because of the use of the `PRIVATE-TOKEN` header, which should be `JOB-TOKEN` for Job Tokens.

Not sure, but I think header `Authorization: Bearer <token>` can handle both?

**Comment by ReazerDev** [comment]
Judging from the age and last update on [this issue](https://gitlab.com/gitlab-org/gitlab/-/issues/225352), the job token probably won't ever be able to access the projects endpoint, atleast not in the near future.

@fgreinacher Should we just skip the permissions check in https://github.com/semantic-release/gitlab/blob/master/lib/verify.js if GITLAB_TOKEN is equal to the CI_JOB_TOKEN?

**Comment by fgreinacher** [comment]
Hm, how would we know that the token is a job token?

**Comment by ReazerDev** [comment]
> Hm, how would we know that the token is a job token?

We'd require the `GITLAB_TOKEN` to be set to the `CI_JOB_TOKEN` in the `.gitlab-ci.yml` like so:

```yaml
variables:
  GITLAB_TOKEN: $CI_JOB_TOKEN
```

and inside of the `verify.js` we'd get both tokens and check if they are equal.

**Comment by OrbisK** [comment]
> > Hm, how would we know that the token is a job token?
> 
> We'd require the `GITLAB_TOKEN` to be set to the `CI_JOB_TOKEN` in the `.gitlab-ci.yml` like so:
> 
> ```yaml
> variables:
>   GITLAB_TOKEN: $CI_JOB_TOKEN
> ```
> 
> and inside of the `verify.js` we'd get both tokens and check if they are equal.

Is there anything against additionally defaulting the GL_TOKEN to the CI_JOB_TOKEN?

**Comment by fgreinacher** [comment]
I don't think we should default the GitLab token to `CI_JOB_TOKEN` as long as it's not enough to support all use cases.

**Comment by toms-place** [comment]
GitLab is now working on allowing the CI_JOB_TOKEN to push to it's own repository. Wouldn't this unblock this issue from getting resolved?
See the current status of GitLabs work [here](https://gitlab.com/gitlab-org/gitlab/-/merge_requests/154111), [here](https://gitlab.com/gitlab-org/gitlab/-/issues/468320#note_2461526493) and [here](https://gitlab.com/gitlab-org/gitlab/-/issues/475705#note_2461781262).

**Comment by SeWieland** [comment]
How is the commit actually created by the plugin?

I have the requirement that all commits must be signed in the repository.

This would easily be possible using the [Commits API](https://docs.gitlab.com/api/commits/#create-a-commit-with-multiple-files-and-actions), as all commits created through REST are signed internally by GitLab itself and are verifiable using the GitLab public key for signing. Also, this might possibly resolve all other issues..

**Comment by travi** [comment]
> How is the commit actually created by the plugin?

also worth noting that no commit is made by default. if you are making a commit during your release, you have opted in to that functionality. we [recommend against it](https://semantic-release.gitbook.io/semantic-release/support/faq#why-is-the-package.jsons-version-not-updated-in-my-repository), so it could be worth considering whether you even actually need to make a commit during the release in the first place.

**Comment by SeWieland** [comment]
This is great news. However, I still get `You are not allowed to push code to this project.`, even if I set `GITLAB_TOKEN: $CI_JOB_TOKEN` => So this is still not possible, even if we do not create any commit? Using the releases API we do not actually need write permissions on the repository, right?

**Comment by texanator** [comment]
[Fine grained permissions](https://docs.gitlab.com/17.11/ci/jobs/fine_grained_permissions/) allow for the CI_JOB_TOKEN to push to its own repo.

**Comment by SeWieland** [comment]
But I don't want the token to have write access on the repository if there is no technical need for it

**Comment by arvest-bjoneson** [comment]
I understand the need to fully support the capabilities of the project. In our case, all I need to do is push a GitLab release (which also creates the tag), which the job token already appears to allow.

This would eliminate a huge headache of having to issue and rotate access tokens. 

Maybe mark the feature as experimental?

**Comment by toms-place** [comment]
@wwuck 
> Will this be enough to allow using `CI_JOB_TOKEN`?

I solved this with CI_JOB_TOKEN and a pipeline trigger, see: https://github.com/semantic-release/gitlab/issues/891#issuecomment-3927050182
